gamemodes/terrortown/gamemode/client/cl_tradio.lua

```diff
@@ -21,92 +20,80 @@ local sound_names = {
-            local snd = row[rk]
-            local rl = rk - 1
-
-            y = rj * m + rj * bh
-            x = rl * m + rl * bw
+---
+-- Calculates and caches the dimensions of the Radio UI.
+-- @realm client
+function TRADIO:CalculateSizes()
+    self.sizes.padding = 10
 
-            local but = vgui.Create("DButton", b)
-            but:SetPos(x, y)
-            but:SetSize(bw, bh)
-            but:SetText(LANG.GetTranslation(sound_names[snd]))
+    self.sizes.heightButton = 45
+    self.sizes.widthButton = 160
 
-            but.snd = snd
-            but.DoClick = ButtonClickPlay
-        end
-    end
+    self.sizes.numWidthButtons = 3
+    self.sizes.numHeightButtons = math.ceil(table.Count(sounds) / 3)
 
-    b:SetSize(bw * hor + m * (hor - 1), bh * ver + m * (ver - 1))
-    b:SetPos(m, 25)
-    b:CenterHorizontal()
+    self.sizes.widthMainArea = self.sizes.widthButton * self.sizes.numWidthButtons
+        + (self.sizes.numWidthButtons - 1) * self.sizes.padding
+    self.sizes.heightMainArea = self.sizes.heightButton * self.sizes.numHeightButtons
+        + (self.sizes.numHeightButtons - 1) * self.sizes.padding
 
-    return b
+    self.sizes.width = self.sizes.widthMainArea + 2 * self.sizes.padding
+    self.sizes.height = self.sizes.heightMainArea
+        + vskin.GetHeaderHeight()
+        + vskin.GetBorderSize()
+        + 2 * self.sizes.padding
 end
 
 ---
--- Creates the traitor radio menu
--- @param Panel parent
--- @return Panel the created DPanel menu
+-- Show the radio UI with all sound buttons. Generates the whole UI.
 -- @realm client
-function TRADIO.CreateMenu(parent)
-    local w, h = parent:GetSize()
-    local client = LocalPlayer()
-
-    local wrap = vgui.Create("DPanel", parent)
-    wrap:SetSize(w, h)
-    wrap:SetPaintBackground(false)
-
-    local dhelp = vgui.Create("DLabel", wrap)
-    dhelp:SetFont("TabLarge")
-    dhelp:SetText(LANG.GetTranslation("radio_help"))
-    dhelp:SetTextColor(COLOR_WHITE)
-
-    if IsValid(client.radio) then
-        CreateSoundBoard(wrap) -- local board
-    elseif client:HasWeapon("weapon_ttt_radio") then
-        dhelp:SetText(LANG.GetTranslation("radio_notplaced"))
+function TRADIO:Toggle(radioEnt)
+    self.radio = radioEnt
+    self.entIndex = tostring(radioEnt:EntIndex())
+
+    self:CalculateSizes()
+
+    -- IF MENU ELEMENT DOES NOT ALREADY EXIST, CREATE IT
+    if IsValid(self.menuFrame) then
+        self.menuFrame:CloseFrame()
+    else
+        self.menuFrame =
+            vguihandler.GenerateFrame(self.sizes.width, self.sizes.height, "radio_name")
     end
 
-    dhelp:SizeToContents()
-    dhelp:SetPos(10, 5)
-    dhelp:CenterHorizontal()
-
-    return wrap
+    self.menuFrame:SetPadding(
+        self.sizes.padding,
+        self.sizes.padding,
+        self.sizes.padding,
+        self.sizes.padding
+    )
+
+    -- any keypress closes the frame
+    self.menuFrame:SetKeyboardInputEnabled(false)
+    self.menuFrame.OnKeyCodePressed = util.BasicKeyHandler
```

Comment:
This means that the UI is closed as soon as the player moves, doesn't it? Not sure if this is the right approach here as it is probably annoying to stand this to interact. With the old shop implementation you were also able to move around while the UI is open

Reply:
I believe line 73 is irrelevant, as I deactivated it in 72. So you can move while running.
I still gotta do some cleanups 😄